src/main/scala/org/clulab/asist/agents/DialogAgentMqtt.scala

```diff
@@ -1,86 +1,249 @@
+    text: String
+  ): Unit = bus.publish(topic, text)
 
   /** Receive a message from the message bus
-   *  @param topic:  The message bus topic where the message was published
-   *  @param json:  A metadata text string
+   * @param topic:  The message bus topic where the message was published
+   * @param text:  A metadata text string, possibly multi-line
    */
   def messageArrived(
     topic: String,
-    json: String
-  ): Unit = topic match {
-    case `topicSubTrial` => trialMessageArrived(json)
-    case _ => json.split("\n").map(line => 
-      allCatch.opt(read[Metadata](line)).map{metadata => 
-        val message = getDialogAgentMessage(
-          source_type,
-          topic,
-          topic,
-          metadata
-        )
-//        dqm.enqueueClassification(publishMessage, message)
+    text: String 
+  ): Unit = {
+
+    // if the queue is empty, there is no aynchronous job in
+    // progress and it is safe to start a new one.
+    val noJobRunning = queue.isEmpty
+
+    // Each line of text becomes a discrete processing job ready to run
+    text.split("\n").map{
+      line => queue.enqueue(BusMessage(topic, line))
+    }
+
+    // start new async job if none are running
+    if(noJobRunning) startJob
+  }
+
+  /** When finished, remove the queue head and start the next job.  */
+  def finishJob: Unit = if(queue.isEmpty) {
+    logger.error("finishJob called with empty queue!")
+  } else {
+    queue.dequeue
+    startJob
+  }
+
+  /* Use the head of the queue as the next job. */
+  def startJob: Unit = if(!queue.isEmpty) queue.head.topic match {
+    case `topicSubTrial` => processTrialMessage(queue.head)
+    case _ => processDialogAgentMessage (queue.head)
+  }  // else all jobs are done.
+
+
+  /** send VersionInfo if we receive a TrialMessage with subtype "start", 
+   * @param input: Message bus traffic with topic and text
+   */
+  def processTrialMessage(input: BusMessage): Unit = try {
+    val tm = read[TrialMessage](input.line)
+    if(tm.msg.sub_type == "start") {
+      val timestamp = Clock.systemUTC.instant.toString
+      writeToMessageBus(topicPubVersionInfo, write(VersionInfo(this, timestamp)))
+      if(withClassifications) resetServer  // enter async execution
+      else finishJob  // no DAC 
+    }
+    else finishJob  // no trial start
+  } catch {
+    case NonFatal(t) => logger.error("Could not parse {}", input.line)
+  } 
+
+  /** Reset the DAC server each time we send a VersionInfo message */
+  def resetServer(): Unit = {
+    logger.info("Resetting DAC server at {}",serverLocation)
+
+    val request = HttpRequest(
+      uri = Uri(s"${serverLocation}/reset-model"),
+      entity = HttpEntity(ContentTypes.`application/json`,"reset-model")
+    )
+    val future: Future[HttpResponse] = Http().singleRequest(request)
+
+    try {
+      val response: HttpResponse = Await.ready(future, 10 seconds).value.get.get
+
+      val futureReply: Future[String] = response.entity.dataBytes
```

Comment:
this line of code seems to be used over and over, maybe it would be good to have it as a method?

maybe some others too?

Reply:
Agreed